crates/eframe/src/web/events.rs

```diff
@@ -590,3 +591,71 @@ pub(crate) fn install_canvas_events(runner_ref: &WebRunner) -> Result<(), JsValu
 
     Ok(())
 }
+
+pub(crate) fn install_resize_observer(runner_ref: &WebRunner) -> Result<(), JsValue> {
+    let closure = Closure::wrap(Box::new({
+        let runner_ref = runner_ref.clone();
+        move |entries: js_sys::Array| {
+            // Only call the wrapped closure if the egui code has not panicked
+            if let Some(mut runner_lock) = runner_ref.try_lock() {
+                runner_lock.needs_repaint.repaint_asap();
+                let canvas = runner_lock.canvas();
+                let (width, height) = get_display_size(&entries).unwrap();
+                canvas.set_width(width);
+                canvas.set_height(height);
+
+                // force an immediate repaint
+                runner_lock.needs_repaint.repaint_asap();
+                paint_if_needed(&mut runner_lock);
+            }
+        }
+    }) as Box<dyn FnMut(js_sys::Array)>);
+
+    let observer = web_sys::ResizeObserver::new(closure.as_ref().unchecked_ref())?;
+    let mut options = web_sys::ResizeObserverOptions::new();
+    options.box_(web_sys::ResizeObserverBoxOptions::ContentBox);
+    observer.observe_with_options(runner_ref.try_lock().unwrap().canvas(), &options);
+    runner_ref.set_resize_observer(observer, closure);
+
+    Ok(())
+}
+
+// Code ported to Rust from:
+// https://webglfundamentals.org/webgl/lessons/webgl-resizing-the-canvas.html
+fn get_display_size(resize_observer_entries: &js_sys::Array) -> Result<(u32, u32), JsValue> {
+    let width;
+    let height;
+    let mut dpr = web_sys::window().unwrap().device_pixel_ratio();
+
+    let entry: web_sys::ResizeObserverEntry = resize_observer_entries.at(0).dyn_into()?;
+    if JsValue::from_str("devicePixelContentBoxSize").js_in(entry.as_ref()) {
+        // NOTE: Only this path gives the correct answer for most browsers.
+        // Unfortunately this doesn't work perfectly everywhere.
```

Comment:
Did you test on any browser that didn't follow this path?

Reply:
I don't remember, will re-test

Reply:
It seems like most of the code is just to support slightly older versions of Safari. Latest Safari doesn't support `devicePixelContentBoxSize` and it seems to work there: https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserverEntry#browser_compatibility

I also commented out all the paths one-by-one to see if they work at all (_except_ the `contentBoxSize` non-array path) and didn't run into any issues.